promote_op() fixes

nalimilan opened the pull request:
I've added systematic tests for `promote_op` and operators on numbers, and they uncovered a few corner cases which either didn't work at all, were type-unstable, or for which `promote_op` did not match the actual type. See commit messages for details.

The first commit comes from https://github.com/JuliaLang/julia/pull/16996 and is needed for all tests to pass, so I moved it here.

Cf. https://github.com/JuliaLang/julia/pull/16988. Fixes JuliaLang/julia#4883.

nalimilan commented:
@JeffBezanson I've found one case where the `promote_op` fallback to `promote_type(S, T)` was used: for operations on Complex. This implied that *(::Matrix{Complex}, Matrix{Complex})` returned a `Matrix{Complex}`, while the rule based on `one` would return a `Matrix{Complex{Int}}`.

While more consistent, the latter has the drawback that `*` no longer works when the matrix contains e.g. `Complex{Float64}` (which is why I changed a test in `linalg/dense.jl`). For example:
```julia
julia> dim=2
2

julia> S=zeros(Complex,dim,dim)
2×2 Array{Complex,2}:
 0+0im  0+0im
 0+0im  0+0im

julia> T=zeros(Complex,dim,dim)
2×2 Array{Complex,2}:
 0+0im  0+0im
 0+0im  0+0im

julia> T[:] = 1
1

julia> z = 2.5 + 1.5im
2.5 + 1.5im

julia> S[1] = z
2.5 + 1.5im

julia> S*T
ERROR: InexactError()
 in convert(::Type{Int64}, ::Float64) at ./int.jl:239
 in matmul2x2!(::Array{Complex{Int64},2}, ::Char, ::Char, ::Array{Complex,2}, ::Array{Complex,2}) at ./linalg/matmul.jl:662
 in generic_matmatmul!(::Array{Complex{Int64},2}, ::Char, ::Char, ::Array{Complex,2}, ::Array{Complex,2}) at ./linalg/matmul.jl:441
 in *(::Array{Complex,2}, ::Array{Complex,2}) at ./linalg/matmul.jl:129
 in eval(::Module, ::Any) at ./boot.jl:231
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```
This problem has been discussed at https://github.com/JuliaLang/julia/issues/4796.

OTOH, the same problem currently exists with reals:
```julia
julia> Real[1 2] * Real[1.5; 2.]
ERROR: InexactError()
 in convert(::Type{Int64}, ::Float64) at ./int.jl:239
 in generic_matvecmul!(::Array{Int64,1}, ::Char, ::Array{Real,2}, ::Array{Real,1}) at ./linalg/matmul.jl:416
 in *(::Array{Real,2}, ::Array{Real,1}) at ./linalg/matmul.jl:84
 in eval(::Module, ::Any) at ./boot.jl:231
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```

Instead of the current hack, one solution would be to change `one` for `Real` (or even `Number`?) to return `1.0::Float64` by default, instead of `1::Int`. Indeed floats are a much better type for representing any real than integers. This is what https://github.com/JuliaLang/julia/pull/6183 did, but was reverted by https://github.com/JuliaLang/julia/issues/6461 since it caused issues. On another open issue (https://github.com/JuliaLang/julia/issues/4808), the consensus seems to be that we need to make `one` an error on abstract types. Should we also make `*` an error?

nalimilan commented on test/broadcast.jl:
```diff
@@ -165,9 +165,9 @@ m = [1:2;]'
 rt = Base.return_types(.+, Tuple{Array{Float64, 3}, Array{Int, 1}})
 @test length(rt) == 1 && rt[1] == Array{Float64, 3}
-rt = Base.return_types(broadcast, Tuple{Function, Array{Float64, 3}, Array{Int, 1}})
+rt = Base.return_types(broadcast, Tuple{typeof(+), Array{Float64, 3}, Array{Int, 1}})
```
This is another place where the return type is assumed to be equal to `promote_type` on the inputs. It breaks with the `promote_op` definition based on `typeof(op(one(R), one(S)))`, since the exact function needs to be specified to call it.

nalimilan commented:
I think I've found a better solution to the abstract types problem from my last comment: in `promote_op`, check whether one of the input types is a supertype of the operation's return type, and if so return the most general type. This is a simple rule, and it even fixes `Real[1 2] * Real[1.5; 2.]`. I've updated the PR with this new approach.

nalimilan commented on base/irrationals.jl:
```diff
@@ -10,6 +10,10 @@ promote_rule{s}(::Type{Irrational{s}}, ::Type{Float32}) = Float32
 promote_rule{s,T<:Number}(::Type{Irrational{s}}, ::Type{T}) = promote_type(Float64,T)
 
+promote_op{S<:Irrational,T<:Irrational}(::Any, ::Type{S}, ::Type{T}) = Float64
+promote_op{S<:Irrational,T<:Number}(::Any, ::Type{S}, ::Type{T}) = promote_type(S, T)
```
`Irrational` looks like a valid case where `promote_op` should rely on `promote_type`: irrationals by definition cannot be represented correctly by any other type, and no other type is closed under operations with irrationals. Therefore, the most appropriate type for the result is the same for all operations, and is obtained using promotion rules.

JeffBezanson replied:
> the most appropriate type for the result is the same for all operations

Only for all operations of a particular, numerical variety. Not for things like converting to string, etc.

nalimilan replied:
Of course. One issue is that the scope of `promote_op` isn't clearly defined: is "op" only for standard operators? For any operation? That's a general fallback anyway.

JeffBezanson replied:
That's true, but `broadcast` uses it for everything. Since `broadcast` doesn't know when `promote_op` is the right thing to use, these fallbacks continue to get in the way.

nalimilan replied:
What would `broadcast` use if we removed the fallbacks? What do you think we should do?

pabloferz replied:
@JeffBezanson commented on #16986 that we might want to have the fallbacks as `Union{}` and use an algorithm similar to the one map uses by dispatching on `Union{}`.

JeffBezanson replied:
If the fallbacks returned `Union{}` or perhaps `Any`, we could dispatch to a version of `broadcast` that uses the `map` algorithm in that case.

nalimilan replied:
But which fallbacks? Should we keep the `Numbers` methods? Or should we restrict them to the combination of numbers and known operators?

JeffBezanson replied:
I don't think it ever makes sense to have a definition of `promote_op` that ignores the function argument --- except one fallback that returns a sentinel value like `Any`.

nalimilan replied:
I general I fully agree. But for the special case of numbers, the use of `one` means that the function argument is taken into account.

Anyway, I only care to the extent that we need a way to get the return type to implement operators on `Nullable` (which are the origin of this investigation).

nalimilan commented:
@JeffBezanson Could you tell me which part of this PR (and of additional fixes at https://github.com/JuliaLang/julia/pull/17114) you would accept, and what we should fix? This is blocking operators support for `Nullable` (https://github.com/JuliaStats/NullableArrays.jl/issues/119), which in turns blocks progress on porting DataFrames to `Nullable`.

JeffBezanson commented:
I would prefer to remove the definitions that ignore `op`. But if that really cannot be done now, you can just go ahead with this PR (tests seem to have timed out for some reason).

pabloferz commented:
I think that in #17114 I managed to leave only `promote_op`s that take `op` into account and anything else will be `Any`. Unless of course I missed something and did something wrong.

nalimilan commented:
I don't think we can get rid of the method for numbers which is based on `one`, but no other method applying to any `op` is added.

I've just picked further changes by @pabloferz, let's see the result.

pabloferz commented on base/float.jl:
```diff
@@ -196,10 +196,10 @@ promote_rule(::Type{Float32}, ::Type{Float16}) = Float32
+promote_op{T<:Union{Float32,Float64}}(::typeof(trunc), ::Type{Unsigned}, ::Type{T}) = UInt
 for f in (ceil, floor, round, trunc)
-    @eval promote_op{Op<:$(typeof(f)),T}(::Op, ::Type{T}, ::Any) = T
+    @eval promote_op{T}(::typeof($f), ::Type{T}, ::Any) = T
```
There's going to be a lot of ambiguities with these changes that are going to be too cumbersome to solve otherwise.

nalimilan replied:
Why? It shouldn't make any difference from the previous syntax.

pabloferz replied:
For the same reason you added versions for `<`, `>`, `==`, etc. parametrised with number. You'll need versions parametrised with `Number` and `Nullable` for `convert`, `ceil`, `parse`, etc.

nalimilan replied:
Have you actually tested it? In my tests, the `Op<:` form has exactly the same ambiguities.

JeffBezanson commented:
Yes, the definitions that do `op(one(T))` are fine.

Don't worry too much about this --- #16622 adds a fairly useful `return_type` function, and #17172 adds a fallback for `broadcast` that will give a good result when `promote_op` returns `Any`.

nalimilan commented:
@JeffBezanson Does that mean we don't need the methods for `convert`, `parse` `round`, etc.? They create a lot of ambiguities with the `Number` methods.

JeffBezanson commented:
Yes, just leave out any tricky definitions. As long as promote_op either returns the right answer or `Any`, everything will work.

JeffBezanson commented:
To be more specific, please delete as many definitions of promote_op as possible.

JeffBezanson commented on base/parse.jl:
```diff
@@ -194,3 +194,5 @@ function parse(str::AbstractString; raise::Bool=true)
     return ex
 end
+
+promote_op{T}(::typeof(parse), ::Type{T}, ::Any) = T
```
This should be removed. Since the first argument to `parse` is a type, it would have to be `Type{Type{T}}`, but that won't be useful very often since an array of types would have element type `DataType`.

nalimilan commented:
OK, I've tried to remove as many definitions as possible from the PR. This fixes the issues with ambiguities. Let's see whether it passes all tests.

JeffBezanson commented:
Could you add the `length.(["foo", "bar"])` and `sin.([1, 2])` tests?

pabloferz commented on base/operators.jl:
```diff
@@ -61,6 +61,13 @@ const ≥ = >=
+    # to fix ambiguities
+    @eval promote_op{R<:Number,S<:Number}(::$(typeof(op)), ::Type{R}, ::Type{S}) =
+        ($(Expr(:meta, :pure)); Bool)
+    @eval promote_op{Op<:$(typeof(op))}(::Op, ::Any, ::Any) = ($(Expr(:meta, :pure)); Bool)
```
If you do this, you don't need the definition above.

nalimilan replied:
I think I do, because the general definition for `Number` applies to all operations, while this one applies only to one operation, for for all input types. So there's still an ambiguity.

nalimilan replied:
Or maybe you're talking about another method?

pabloferz replied:
I was referring to this one. But you are right.

pabloferz commented on base/promotion.jl:
```diff
@@ -222,9 +222,8 @@ minmax(x::Real, y::Real) = minmax(promote(x, y)...)
 #   promote_op{R<:MyType,S<:MyType}(::typeof(*), ::Type{R}, ::Type{S}) = MyType{multype(R,S)}
 promote_op(::Any)    = (@_pure_meta; Bottom)
-promote_op(::Any, T) = (@_pure_meta; T)
-promote_op{T}(::Type{T}, ::Any) = (@_pure_meta; T)
```
I believe this is still useful for doing things like `Float64.([1,2,3])`

nalimilan replied:
Seems to work without it. :-)

JeffBezanson replied:
A potential issue to be aware of here is that e.g. `Real(1+0im)` returns an `Int`, so we don't necessarily want to say the result is `Real`.

nalimilan replied:
Why not? Sounds like a case where you might want to store non-`Int` values in the resulting array:
```julia
julia> x = Real.([1+0im])
1-element Array{Real,1}:
 1

julia> x[1] = 1.5
1.5
```

pabloferz replied:
`promote_op{T}(::Type{T}, ::Any) = (@_pure_meta; T)` was added in #16179 without things like `Float32.([1,2,3])` will return `Any` and that will fallback to the new slow broadcast algorithm. I think it it's convenient to keep it. Of course, I'll leave it for you to consider it.

nalimilan replied:
I'm fine with adding it back, but then where should we place the limit? It's annoying that the fallback is slow in any case.

pabloferz replied:
I just noticed that you have `promote_op{R,S<:Number}(::Type{R}, ::Type{S}) = (@_pure_meta; R)` above so that will get picked instead if you remove this. So I guess there is no problem leaving this out.

JeffBezanson replied:
The fallback is not necessarily that slow --- in cases we can infer, there is no significant slowdown at all. However there is a chance we will lose some SIMD optimizations (though I haven't analyzed it in detail), so we should keep using the old approach for +, -, *, etc.

I'm also ok with adding back this definition.

nalimilan replied:
I've added it back, that's an important special case and it doesn't create any issues.

nalimilan commented:
Will do, but the former doesn't return the most restricted type it could. Any idea why?
```julia
julia> length.(["foo", "bar"])
2-element Array{Any,1}:
 3
 3

```

nalimilan commented:
There's a failure in the `broadcast` test. Why doesn't this return `Array{Float64, 1}`? Yet type inference and `promote_op` work fine here.
```julia
julia> rt = Base.return_types(broadcast, Tuple{Float64, Array{Int, 1}})
1-element Array{Any,1}:
 Array{T,N}

julia> @inferred broadcast(Float64, [1,2])
2-element Array{Float64,1}:
 1.0
 2.0


```

nalimilan commented:
Nevermind, this needs to be `Type{Float64}` of course.

nalimilan commented:
A more serious one:
```julia
julia> broadcast(tuple, [1 2 3], ["a", "b", "c"])
3×3 Array{Any,2}:
 (1,"a")  (2,"a")  (3,"a")
 (1,"b")  (2,"b")  (3,"b")
 (1,"c")  (2,"c")  (3,"c")
```

It used to return `Matrix{Tuple{Int,String}}`. I wonder how my changes can have affected this...

nalimilan commented:
@JeffBezanson There's another problem: there are other places where `promote_op` is used without the `broadcast` type algorithm. For example, `-(::AbstractArray, ::AbstractArray)` relies on it to choose the element type of the result. The [removal of the fallback definition based on `promote_type`](https://github.com/JuliaLang/julia/pull/16995/files#diff-e7776a0970cdf71c5c96557db2986f86L227) means that we now return an `Array{Any}` by default for non-Numbers. This already triggers a failure in the `resolve` test, which something similar to `x = [Base.Pkg.Resolve.MaxSum.FieldValues.FieldValue(0)]; abs(x-x)`.

I guess the solution is to use `broadcast(-, ...)` to implement this method?

JeffBezanson commented:
The promote_op method calling promote_type could be restored for `Union{typeof(+), typeof(-), ...}` as the type of `op`.

nalimilan commented:
> The promote_op method calling promote_type could be restored for Union{typeof(+), typeof(-), ...} as the type of op.

I don't really like that solution either, since that rule isn't correct in many cases: `-(::Date, ::Date)` is an example for which we have a custom `promote_op` rule in Base. I'd rather return `Any` and have an algorithm find the most appropriate type (maybe using return type declarations? that would be a nicer API than `promote_op`).

So, as regards the `resolve` failure in Base, I've just added `promote_op` methods for `+` and `-`, which are the only supported operations. We can easily revise this later when the whole system will have been improved.

pabloferz commented on base/number.jl:
```diff
@@ -63,4 +63,16 @@ zero{T<:Number}(::Type{T}) = convert(T,0)
+    # preserve the most general (abstract) type when possible
+    return isleaftype(T) ? S : typejoin(S, T)
+end
+function promote_op{R<:Number,S<:Number}(op, ::Type{R}, ::Type{S})
```
Maybe we can write

```
x = R <: Irrational ? 1.0 : one(R)
y = S <: Irrational ? 1.0 : one(S)
T = typeof(op(x, y))
```

and remove the ones for `Irrational` above?

nalimilan replied:
Yeah, I thought about that possibility too. In general though, I think it's considered more Julian to dispatch on types.

That solution would be quite appealing to avoid ambiguities with the methods for comparison operators, but we still need to take care of ambiguities with the `Nullable` method. So I'd rather get rid of the comparison operators method altogether. That fits with @JeffBezanson's request. (But currently it fails the tests, not sure why yet.)

pabloferz replied:
I now dispatching in type is more Julian. The problem here, if I understand Jeff right, is that any method that ignores op, that is `promote_op(op, ...)` should preferably return `Any`, and that we should have as fewer such methods as possible (but also the as fewer as possible of the ones that take op into account).

The methods that have `Number` types as arguments (other than op) are a special case since we used in other places beside broadcast. The methods with `Nullable` types are probably also important on its own.

The comparison methods are failing because there is no `isless` for `Complex` and other numerical types.

nalimilan replied:
> I'm not sure why the comparison methods can't be inferred with this definition, that's why I had the comparison ones.

I suspect there's a type-instability somewhere. These tests already caught one in a corner case.

nalimilan replied:
Found a minimum reproducer: `Base.Test.@inferred <(one(Rational{Int32}), one(Float64))` (only happens on 32-bit)

pabloferz commented on base/irrationals.jl:
```diff
@@ -10,6 +10,13 @@ promote_rule{s}(::Type{Irrational{s}}, ::Type{Float32}) = Float32
 promote_rule{s,t}(::Type{Irrational{s}}, ::Type{Irrational{t}}) = Float64
 promote_rule{s,T<:Number}(::Type{Irrational{s}}, ::Type{T}) = promote_type(Float64,T)
 
+promote_op{S<:Irrational,T<:Irrational}(::Any, ::Type{S}, ::Type{T}) =
```
This is missing op as first argument

nalimilan replied:
Good catch. Will fix, but the CI hang doesn't seem related...

nalimilan committed:
Fix type instability in decompose(::Float64) on 32-bit

Need at least 64-bit to store the result of Float64 decomposition.
(A very good use case for return type declarations.)

pabloferz commented on base/intfuncs.jl:
```diff
@@ -198,7 +198,7 @@ end
     if b < 0
-        d = ndigits0znb(signed(n), b)
+        d = ndigits0znb(n, b)
     else
```
CI has been hanging all along because of this (see https://github.com/JuliaLang/julia/blob/master/test/intfuncs.jl#L78)

nalimilan replied:
Thank you so much for catching it! How did you found out that was the culprit?

pabloferz replied:
I looked in `test/choosetests.jl` to see if there was some of the tests missing in the log.

nalimilan replied:
I had tried that too, but I guess it didn't do it systematically enough. Adding a timeout on individual tests would indeed make this much clearer...

nalimilan commented:
> add some @inferred tests for this?

The new tests cover it already, that's why they were failing.

pabloferz commented:
AV failure seems to be #16091

nalimilan commented:
Yes, I've restarted it hoping we'll be more lucky this time.

Good news: the Travis CI build passed! Is there anything left to change? I'd be in favor of merging if it's globally OK, and possibly tweak details in subsequent PRs.

JeffBezanson commented:
Thanks for persisting on this! Could you squash some of the commits? In particular would be good to keep the ones that fix general issues separate, and squash together the promote_op changes.

tkelman commented on base/intfuncs.jl:
```diff
@@ -198,7 +198,7 @@ end
     d = 0
     if b < 0
-        d = ndigits0znb(n, b)
+        d = ndigits0znb(signed(n), b)
```
is this still right if it wraps to negative?

nalimilan replied:
This change actually cancels the one I incorrectly made in a previous commit. So after squashing nothing changes here.

tkelman replied:
I guess consider that a question w.r.t. master then?

nalimilan replied:
I think so, since after removing `signed` the unsigned wrapping triggered a failure. So I guess this means with `signed`, the wrapping is correct (since the test works). But you'll have to find somebody who knows this code to get an actual answer. :-)

nalimilan committed:
Fix various corner cases with operators on Numbers

These only affected particular type combinations, with either
type instabilities which affected inference of return type
or plain failures. Tests are included in the next commit.

nalimilan committed:
promote_op() improvements

Use common promote_op() based on one() for all Number types:
this fixes promote_op(+, ::Bool), which returned Int, and
promote_op(==, ::Complex, ::Complex), which returned Complex{Bool}.
Add fallback definitions returning Any so that broadcast() computes
the appropriate return type, instead of using promote_type() which
is often wrong.

This fixes broadcast() when the return type is different from the input
type. Add systematic tests for operators and their return types.

nalimilan commented:
Squashed and rebased, tests still pass!

nalimilan commented on base/promotion.jl:
```diff
@@ -222,10 +222,8 @@ minmax(x::Real, y::Real) = minmax(promote(x, y)...)
+promote_op(::Any, ::Any, ::Any...) = (@_pure_meta; Any)
 promote_op{T}(::Type{T}, ::Any) = (@_pure_meta; T)
-promote_op{R,S}(::Any, ::Type{R}, ::Type{S}) = (@_pure_meta; promote_type(R, S))
-promote_op(op, T, S, U, V...) = (@_pure_meta; promote_op(op, T, promote_op(op, S, U, V...)))
```
I've removed this method when squashing since it doesn't make sense to use the output type as an input type now that the fallback to `promote_type` has been removed. I've replaced it with `promote_op(::Any, ::Any, ::Any...)` above.

pabloferz replied:
Makes sense :+1:. Isn't this just missing `promote_op(::Any, ::Any) =  Any`?

nalimilan replied:
No, because `::Any...` can also match no argument. This is an annoying detail with current varargs.

nalimilan commented on base/complex.jl:
```diff
@@ -461,7 +461,7 @@ function ^{T<:AbstractFloat}(z::Complex{T}, p::Complex{T})
         zr, zi = reim(z)
         x = (zr-zi)*(zr+zi)
-        y = 2zr*zi
+        y = T(2)*zr*zi
```
Because of the type instability it creates for `Float16`:
```julia
julia> typeof(2*Float16(1.0))
Float32
```

Maybe better write it as `T(2*zr*zi)`?

(I wonder whether that promotion exception for `Float16` is really a good idea...)

stevengj replied:
Yes, I think `T(2*zr*zi)` seems like a better way to do it.

nalimilan replied:
See https://github.com/JuliaLang/julia/pull/17259.
